controller/system_restore_controller.go

```diff
@@ -0,0 +1,397 @@
+package controller
+
+import (
+	"fmt"
+	"reflect"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/kubernetes/pkg/controller"
+
+	batchv1 "k8s.io/api/batch/v1"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	clientset "k8s.io/client-go/kubernetes"
+	v1core "k8s.io/client-go/kubernetes/typed/core/v1"
+
+	"github.com/longhorn/longhorn-manager/datastore"
+	"github.com/longhorn/longhorn-manager/engineapi"
+	"github.com/longhorn/longhorn-manager/types"
+
+	longhorn "github.com/longhorn/longhorn-manager/k8s/pkg/apis/longhorn/v1beta2"
+)
+
+const (
+	SystemRestoreControllerName = "longhorn-system-restore"
+
+	RestoreJobBackoffLimit = 3
+)
+
+type SystemRestoreController struct {
+	*baseController
+
+	// which namespace controller is running with
+	namespace string
+	// use as the OwnerID of the controller
+	controllerID string
+
+	kubeClient    clientset.Interface
+	eventRecorder record.EventRecorder
+
+	ds *datastore.DataStore
+
+	cacheSyncs []cache.InformerSynced
+}
+
+func NewSystemRestoreController(
+	logger logrus.FieldLogger,
+	ds *datastore.DataStore,
+	scheme *runtime.Scheme,
+	kubeClient clientset.Interface,
+	namespace string,
+	controllerID string) *SystemRestoreController {
+
+	eventBroadcaster := record.NewBroadcaster()
+	eventBroadcaster.StartLogging(logrus.Infof)
+	// TODO: remove the wrapper when every clients have moved to use the clientset.
+	eventBroadcaster.StartRecordingToSink(&v1core.EventSinkImpl{
+		Interface: v1core.New(kubeClient.CoreV1().RESTClient()).Events(""),
+	})
+
+	c := &SystemRestoreController{
+		baseController: newBaseController(SystemRestoreControllerName, logger),
+
+		namespace:    namespace,
+		controllerID: controllerID,
+
+		ds: ds,
+
+		kubeClient:    kubeClient,
+		eventRecorder: eventBroadcaster.NewRecorder(scheme, corev1.EventSource{Component: SystemRestoreControllerName + "-controller"}),
+	}
+
+	ds.SystemRestoreInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
```

Comment:
Do we need periodical queue resync? or just rely on the resource update only

Reply:
Removed periodical queue resync.